sklearn/metrics/pairwise.py

```diff
@@ -829,6 +829,232 @@ def cosine_distances(X, Y=None):
+        # avoid division by zero when all values in the column are the same
+        ranges[ranges == 0] = 1
+
+        # check if the data is pre-scaled when scale=False
+        if not process_scale and (np.min(min) < 0 or np.max(max) > 1):
+            raise ValueError("Input data is not scaled between 0 and 1.")
+
+    D = np.zeros((X.shape[0], Y.shape[0]), dtype=np.float)
+
+    for i in range(X.shape[0]):
+        j_start = i
+
+        # For non square results
+        if X.shape[0] != Y.shape[0] or X is not Y:
+            j_start = 0
+
+        # Makes the comparisson for np.nan for arrays with dtype=np.object,
+        # this is necessary as some deployments returns True for
+        # np.nan == np.nan
+        cat_nan_cols = (_object_dtype_isnan(X[i, cat_mask]) |
+                        _object_dtype_isnan(Y[j_start:, cat_mask]))
+
+        # Calculates the similarities for categorical columns
+        cat_dists = ((X[i, cat_mask] != Y[j_start:, cat_mask]) | cat_nan_cols)
+        # Calculates the Manhattan distances for numerical columns
+        num_dists = abs(X[i, num_mask] -
+                        Y[j_start:, num_mask]) / ranges
+
+        # Calculates the number of non missing columns
+        non_missing = X.shape[1] - (cat_nan_cols.sum(axis=1) +
+                                    _object_dtype_isnan(num_dists).sum(axis=1)
+                                    .astype(np.float32))
+
+        # This is to avoid ZeroDivisionError
+        non_missing[non_missing == 0] = np.nan
+
+        # Gets the final results
+        total = np.sum(cat_dists, axis=1) + np.sum(num_dists, axis=1)
+
+        results = total / non_missing
+
+        D[i, j_start:] = results
+        if X is Y:
+            D[i:, j_start] = results
+
+    return D
+
+
+def _detect_categorical_features(X, categorical_features=None):
+    """Identifies the numerical and non-numerical (categorical) columns
+    of an array.
+
+    Parameters
+    ----------
+    X : array-like, or pandas.DataFrame, shape (n_samples, n_features)
+
+    categorical_features : array-like, optional, shape (n_features)
+        Indicates with True/False whether a column is a categorical attribute.
+
+        Alternatively, the categorical_features array can be represented only
+        with the numerical indexes of the categorical attribtes.
+
+        If the categorical_features array is None, they will be automatically
+        detected in X. Numerical columns are identified as a subtype of
+        np.number, whilist categorical columns are not a subtype of np.number.
+
+    Returns
+    -------
+    categorical_features_mask : ndarray, shape (n_features)
+
+    """
+    # Automatic detection of categorical features
+    if categorical_features is None:
+        categorical_features = np.zeros(np.shape(X)[1], dtype=bool)
+
+        def detect_cat(x):
+            if not np.isnan(x):
+                if np.issubdtype(type(x), np.number):
+                    raise ValueError(False)
```

Comment:
This looks like a very unconventional way of providing control flow and passing values around. Why are we using exceptions rather than return values here?

Reply:
Okay. I see that we're applying pyufunc to check each element individually, and using exceptions to abort as soon as we have a non-NaN. This logic is very unclear from your code, and I see no benefit in doing it this way rather than an explicit python loop over elements, or something more functional-style:
```py
non_nan_values = itertools.dropwhile(np.isnan, X[:, col])
try:
    value = next(non_nan_values)
except StopIteration:
    TODO: handle case when all values are NaN

TODO: determine type from value
```